File: src/api/git/branches.ts
import { invoke } from '@tauri-apps/api/core';

export interface BranchInfo {
  name: string;
  shorthand: string;
  is_head: boolean;
  is_remote: boolean;
  upstream?: string;
  ahead?: number;
  behind?: number;
  target_commit?: string;
}

export interface BranchComparisonFile {
  path: string;
  status: string;
}

export interface BranchComparison {
  ahead: number;
  behind: number;
  files: BranchComparisonFile[];
}

export async function listBranches(): Promise<BranchInfo[]> {
  return await invoke('list_branches');
}

export async function compareBranches(branchA: string, branchB: string): Promise<BranchComparison> {
  return await invoke('compare_branches', { branchA, branchB });
}

export async function checkoutBranch(branchName: string, force: boolean = false): Promise<void> {
  await invoke('checkout_branch', { branchName, force });
}

export async function createBranch(branchName: string, startPoint?: string): Promise<void> {
  await invoke('create_branch', { branchName, startPoint: startPoint || null });
}

export async function deleteBranch(branchName: string): Promise<void> {
  await invoke('delete_branch', { branchName });
}

export async function renameBranch(oldName: string, newName: string): Promise<void> {
  await invoke('rename_branch', { oldName, newName });
}

export async function deleteRemoteBranch(remoteName: string, branchName: string): Promise<string> {
  return await invoke('delete_remote_branch', { remoteName, branchName });
}

export async function branchFromStash(index: number, branchName: string): Promise<void> {
  await invoke('branch_from_stash', { index, branchName });
}

File: src-tauri/src/commands/branches.rs
use crate::git;
use crate::state::AppState;
use tauri::State;

#[tauri::command]
pub fn list_branches(state: State<'_, AppState>) -> crate::error::Result<Vec<git::BranchInfo>> {
    let repo_path = state.repo_path.lock().unwrap();
    if let Some(p) = repo_path.as_ref() {
        git::list_branches(p)
    } else {
        Err(crate::error::GitError::CommandFailed(
            "No repository is currently open.".to_string(),
        ))
    }
}

#[tauri::command]
pub fn checkout_branch(
    branch_name: String,
    force: bool,
    state: State<'_, AppState>,
) -> crate::error::Result<()> {
    let repo_path = state.repo_path.lock().unwrap();
    if let Some(p) = repo_path.as_ref() {
        git::checkout_branch(p, branch_name, force)
    } else {
        Err(crate::error::GitError::CommandFailed(
            "No repository is currently open.".to_string(),
        ))
    }
}

#[tauri::command]
pub fn create_branch(
    branch_name: String,
    start_point: Option<String>,
    state: State<'_, AppState>,
) -> crate::error::Result<()> {
    let repo_path = state.repo_path.lock().unwrap();
    if let Some(p) = repo_path.as_ref() {
        git::create_branch(p, branch_name, start_point)
    } else {
        Err(crate::error::GitError::CommandFailed(
            "No repository is currently open.".to_string(),
        ))
    }
}

#[tauri::command]
pub fn delete_branch(branch_name: String, state: State<'_, AppState>) -> crate::error::Result<()> {
    let repo_path = state.repo_path.lock().unwrap();
    if let Some(p) = repo_path.as_ref() {
        git::delete_branch(p, branch_name)
    } else {
        Err(crate::error::GitError::CommandFailed(
            "No repository is currently open.".to_string(),
        ))
    }
}

#[tauri::command]
pub fn rename_branch(
    old_name: String,
    new_name: String,
    state: State<'_, AppState>,
) -> crate::error::Result<()> {
    let repo_path = state.repo_path.lock().unwrap();
    if let Some(p) = repo_path.as_ref() {
        git::rename_branch(p, old_name, new_name)
    } else {
        Err(crate::error::GitError::CommandFailed(
            "No repository is currently open.".to_string(),
        ))
    }
}

#[tauri::command]
pub fn delete_remote_branch(
    remote_name: String,
    branch_name: String,
    state: State<'_, AppState>,
) -> crate::error::Result<String> {
    let repo_path = state.repo_path.lock().unwrap();
    if let Some(p) = repo_path.as_ref() {
        git::delete_remote_branch(p, remote_name, branch_name)
    } else {
        Err(crate::error::GitError::CommandFailed(
            "No repository is currently open.".to_string(),
        ))
    }
}

#[tauri::command]
pub fn compare_branches(
    branch_a: String,
    branch_b: String,
    state: State<'_, AppState>,
) -> crate::error::Result<git::BranchComparison> {
    let repo_path = state.repo_path.lock().unwrap();
    if let Some(p) = repo_path.as_ref() {
        git::compare_branches(p, branch_a, branch_b)
    } else {
        Err(crate::error::GitError::CommandFailed(
            "No repository is currently open.".to_string(),
        ))
    }
}

#[tauri::command]
pub fn branch_from_stash(
    index: usize,
    branch_name: String,
    state: State<'_, AppState>,
) -> crate::error::Result<()> {
    let repo_path = state.repo_path.lock().unwrap();
    if let Some(p) = repo_path.as_ref() {
        git::branch_from_stash(p, index, branch_name)
    } else {
        Err(crate::error::GitError::CommandFailed(
            "No repository is currently open.".to_string(),
        ))
    }
}
